### PR TITLE
Make wxRegEx copyable

### PR DESCRIPTION
**Pull request by AliKet**
By making the handle `wxRegExImpl` a ref-counted class, `wxRegEx` now have value semantic.

### CODE REVIEW

**Comment by AliKet** [comment]
Hi @vadz any objection on this ?
This will make wxRegexTextValidator impl. simpler and more cleaner.

File: src/common/regex.cpp [hunk 1]
```diff
@@ -627,22 +627,49 @@ void wxRegEx::Init()
...
+    if ( m_impl )
+        m_impl->DecRef();
+
+    m_impl = new wxRegExImpl;
 
     if ( !m_impl->Compile(expr, flags) )
     {
         // error message already given in wxRegExImpl::Compile
-        wxDELETE(m_impl);
+        m_impl->DecRef();
```
Comment (vadz) [inline src/common/regex.cpp]: Must reset `m_impl` to `NULL` here!

**Review by vadz — COMMENTED** [review]
At the implementation level, we probably should derive (privately) `wxRegEx` from `wxObjectDataPtr<wxRegExImpl>` and then just add using declarations to make `operator=` public.

But I'm more worried about the semantics as it's not really expected that the assert in code like this will fail:
```cpp
wxRegEx r1("foo");
wxRegEx r2 = r1;
r2.Compile("bar");
assert( r1.Matches("foo") );
```
because calling `r2.Compile()` modifies the RE used by `r1`.

If we add ref counting, we need to also implement COW i.e. both `Compile()` and `Matches()` ought to create a new `wxRegExImpl`, which means that we need to store the text of the regex, but it's probably not too bad as REs tend to be relatively short.

It still needs to be done, however, and we need to add tests checking that examples like the above one work as expected.

And, of course, documentation would need to be updated to mention that this class is now copyable.

To summarize, while I agree that it would be nice if you could copy `wxRegEx` objects, in the current form this is too dangerous to apply, sorry.

File: src/common/regex.cpp (hunk 1 above)
Comment (AliKet) [reply src/common/regex.cpp]: Yes m_impl should be reset here.

**Comment by AliKet** [comment]
> ```
> wxRegEx r1("foo");
> wxRegEx r2 = r1;
> r2.Compile("bar");
> assert( r1.Matches("foo") );
> ```

No, this does not fail (tested it).
`Compile()` and the **assignment operator** both implement the COW idiom correctly, You can review this and make sure of it.

For `Matches()` there is no need for a new `wxRegExImpl`

**Comment by vadz** [comment]
Sorry, I just don't understand how is this possible: I don't see any COW implementation in this PR, there is no "unsharing" of `wxRegExImpl`. What am I missing?

**Comment by AliKet** [comment]
I am sorry too for not elaborate further.

The COW am talking about here is touching `m_RegEx` (& `m_isCompiled`) only, because there is no point on making `wxRegExImpl` ref-counted in the first place if we have to consider `m_Matches` (& `m_nMatches`) too which now serve(s) as buffer meaningful only to the last `wxRegEx` object having made a successful call to `Matches()`, and the results of `GetMatch()` & `GetMatchCount()` should be taken or saved a.s.a.p because a successful call to `Matches()` by a (`wxRegEx`) object will invalidate a previous successful call to `Matches()` made by another (`wxRegEx`) object (assuming they both share the same `wxRegExImpl`).

What I mean is better illustrated with a piece of code:

```
wxRegEx r1("foo"); // valid RE
wxRegEx r2; // invalid RE (need to call Compile())

r2 = r1; // now r1 and r2 both share the same wxRegExImpl

assert( r1.GetMatchCount() == 0 );  // need to call Matches() first
assert( r2.GetMatchCount() == 0 ); // idem

assert( !r1.Matches("bar") );

assert( r2.Matches("foo") );
assert( r2.GetMatchCount() == 1 );
assert( r1.GetMatchCount() == 0 ); // the match was made by r2 not by r1.
assert( r1.Matches("foo") ); // will invalidate r2.Matches() above.
assert( r1.GetMatchCount() == 1 );
assert( r2.GetMatchCount() == 0 );

r2.Compile(Invalid_RE); // recompile with invalid RE
assert( !r2.IsValid() ); // now become invalid
assert( r1.IsValid() ); // but r1 is still valid and unaffected by r2 being invalidated.

r2 = r1; // make r2 valid again
r2.Compile("bar"); // r1 and r2 each has its own wxRegExImpl

assert( r1.Matches("foo") );
assert( r2.Matches("bar") ); // will not invalidate r1.Matches()
assert( r1.GetMatchCount() == 1 );
assert( r2.GetMatchCount() == 1 );
```

please take your time to review this, and thank you in advance for your patience and help.

AliKet.

**Comment by AliKet** [comment]
Played with this and it worked as expected

```
void MyFrame::OnTest(wxCommandEvent& WXUNUSED(event))
{
    wxRegEx r1("wx(GTK|MSW|OSX){1}( TK)?"); // valid RE
    wxRegEx r2; // invalid RE (need to call Compile())

    r2 = r1; // now r1 and r2 both share the same wxRegExImpl

    wxASSERT( r1.GetMatchCount() == 0 ); // need to call Matches() first
    wxASSERT( r2.GetMatchCount() == 0 ); // idem

    wxASSERT( !r1.Matches("wxUNIV") );

    wxASSERT( r2.Matches("wxMSW") );
    wxASSERT( r2.GetMatchCount() == 3 );
    wxASSERT( r1.GetMatchCount() == 0 ); // the match was made by r2 not by r1.
    wxASSERT( r1.Matches("wxOSX") ); // will invalidate r2.Matches() above.
    wxASSERT( r1.GetMatchCount() == 3 );
    wxASSERT( r2.GetMatchCount() == 0 );

    {
        wxLogNull noLog;
        r2.Compile("[[[~"); // recompile with invalid RE
    }

    wxASSERT( !r2.IsValid() ); // now becomes invalid
    wxASSERT( r1.IsValid() ); // but r1 is still valid and unaffected by r2 being invalidated.

    r2 = r1; // make r2 valid again
    r2.Compile("wxQt"); // r1 and r2 each has its own wxRegExImpl

    const wxString gtkTK = "wxGTK TK";

    wxASSERT( r1.Matches(gtkTK) );
    wxASSERT( r2.Matches("wxQt") ); // will not invalidate r1.Matches()
    wxASSERT( r1.GetMatchCount() == 3 );
    wxASSERT( r2.GetMatchCount() == 1 );

    r2 = r1;

    wxASSERT( r1.GetMatchCount() == 3 );
    wxASSERT( r2.GetMatchCount() == 0 );

    wxString gtkToolKit = gtkTK;

    r2.Replace(&gtkToolKit, "wx\\1 ToolKit");

    wxASSERT( gtkToolKit == "wxGTK ToolKit" );

    wxASSERT( r1.GetMatchCount() == 3 );
    wxLogMessage("%s", r1.GetMatch(gtkTK, 2)); // will report TK
}
```

**Comment by vadz** [comment]
> Played with this and it worked as expected

Sorry, I couldn't look at it in details yet, but how is this "as expected":
 
>     wxASSERT( r2.Matches("wxMSW") );
>     wxASSERT( r2.GetMatchCount() == 3 );
>     wxASSERT( r1.GetMatchCount() == 0 ); // the match was made by r2 not by r1.
>     wxASSERT( r1.Matches("wxOSX") ); // will invalidate r2.Matches() above.
>     wxASSERT( r1.GetMatchCount() == 3 );
>     wxASSERT( r2.GetMatchCount() == 0 );

? Surely calling `r1.Matches()`, or doing anything else with `r1`, must not change `r2.GetMatchCount()` result! This is exactly the problem I was speaking of. You can't reuse the same matches between different objects, this is totally unexpected and introduces extremely spooky action at a distance. This must be fixed by unsharing the matches objects whenever they're modified!

**Comment by AliKet** [comment]
Does this looks correct ?

```
void MyFrame::OnTest(wxCommandEvent& WXUNUSED(event))
{
    wxRegEx r1("wx(GTK|MSW|OSX){1}( TK)?"); // valid RE
    wxRegEx r2; // invalid RE (need to call Compile())

    // The number of bracketed subexpressions plus one for the expression itself
    const size_t expectedMatchCount = 3;

    r2 = r1; // now r1 and r2 both share the same wxRegExImpl

    wxASSERT( r1.GetMatchCount() == expectedMatchCount );
    wxASSERT( r2.GetMatchCount() == expectedMatchCount );

    const wxString mswPort = "wxMSW port";

    wxASSERT( !r1.Matches("wxUNIV") );
    wxASSERT( r2.Matches(mswPort) );
    wxASSERT( r2.GetMatch(mswPort) == "wxMSW" );
    wxASSERT( r1.GetMatch(mswPort) != "wxMSW" ); // r1 was matched against "wxUNIV"
    wxASSERT( r1.Matches("wxOSX") );
    wxASSERT( r1.GetMatch(mswPort) = "wxOSX" );
    wxASSERT( r2.GetMatch(mswPort) == "wxMSW" ); // still valid

    {
        wxLogNull noLog;
        r2.Compile("[[[~"); // recompile with invalid RE
    }

    wxASSERT( !r2.IsValid() ); // now becomes invalid
    wxASSERT( r1.IsValid() ); // but r1 is still valid and unaffected by r2 being invalidated.

    r2 = r1; // make r2 valid again
    r2.Compile("wxQt"); // r1 and r2 each has its own wxRegExImpl

    const wxString gtkTK = "wxGTK TK";

    wxASSERT( r1.Matches(gtkTK) );
    wxASSERT( r2.Matches("wxQt") );
    wxASSERT( r1.GetMatchCount() == 3 );
    wxASSERT( r2.GetMatchCount() == 1 );

    r2 = r1; // reassign r1 to r2

    wxASSERT( r1.GetMatchCount() == 3 );
    wxASSERT( r2.GetMatchCount() == 3 );

    wxASSERT( r1.GetMatch(gtkTK, 2) == " TK" ); // 
    wxASSERT( r2.GetMatch(gtkTK, 2) == " TK" ); // 

    wxString gtkToolKit = gtkTK;

    r2.Replace(&gtkToolKit, "wx\\1 ToolKit");

    wxASSERT( gtkToolKit == "wxGTK ToolKit" );

    wxASSERT( r1.GetMatchCount() == 3 );
    wxLogStatus("%s", r1.GetMatch(gtkTK, 2)); // will report TK
}
```

If so I'll create another PR with the new implementation.

Thank you!

**Comment by AliKet** [comment]
Here is the new [PR](https://github.com/wxWidgets/wxWidgets/pull/1209)

TIA!
